# Missing Features

***

The following code was used to obtain the audio features of the songs that were missing from favFeatures; Using [Spotipy](https://spotipy.readthedocs.io) - a lightweight python library for the Spotify Web API.

## Data

- fav is a dataframe of my favourite songs, which were songs that I've listened to more than or equal to 9 times.
- favFeatures is a dataframe of audio features for the corresponding track in fav dataframe. 

## Credits

https://towardsdatascience.com/how-to-create-large-music-datasets-using-spotipy-40e7242cc6a6

In [1]:
#import libraries
import pandas as pd 
import spotipy 
import numpy as np

#spotipy authorization
sp = spotipy.Spotify() 
from spotipy.oauth2 import SpotifyClientCredentials 

cid="CLIENT_ID" 
secret="CLIENT_SECRET"
username="USERNAME"

client_credentials_manager = SpotifyClientCredentials(client_id=cid, client_secret=secret) 
sp = spotipy.Spotify(client_credentials_manager=client_credentials_manager) 

sp.trace=False 

In [2]:
#load in fav and favFeatures dataset
fav = pd.read_csv('fav.csv')
favFeatures = pd.read_csv('favFeatures.csv')

In [3]:
fav

artist                                              track  \
0        Ty Dolla $ign                       Purple Emoji (feat. J. Cole)   
1           Dreamville  Sacrifices (with EARTHGANG & J. Cole feat. Smi...   
2              J. Cole                                     everybody dies   
3              J. Cole                                             Change   
4              J. Cole                        t h e . c l i m b . b a c k   
..                 ...                                                ...   
186  The Neighbourhood                                    Sweater Weather   
187         Dean Lewis                                         Be Alright   
188           Doja Cat                                             Say So   
189     Arizona Zervas                                            ROXANNE   
190         Ed Sheeran                    Beautiful People (feat. Khalid)   

     count  
0       75  
1       73  
2       73  
3       70  
4       69  
..     ...  
186      9  
187      9  
188      9  
189      9  
190      9  

[191 rows x 3 columns]

In [4]:
favFeatures

id                                track  popularity  \
0   4u7EnebtmKWzUH433cf5Qv  Bohemian Rhapsody - Remastered 2011          83   
1   4BM8yJ0PzBi2ZewpMTOxtx                             Fernando          68   
2   5qqabIl2vWzo9ApSC317sa              Wonderwall - Remastered          79   
3   3AJwUDP919kvQ9QcozQPxg                               Yellow          86   
4   0I3q5fE6wg7LIfHGngUTnV                          Ms. Jackson          82   
..                     ...                                  ...         ...   
63  4hWgC5BV5M7EtTleKVXNxw                        Playing Games          68   
64  4RHyIzFRht2z17XeHB9dsF                  Through and Through          69   
65  3xaugmCyXrVkrDTXbFkMW3                            Sanctuary          72   
66  63bAGRSSX2V1hhPSP2NpBC                           North Face          69   
67  3f50PY0yOtMHo9BgCX9WRJ                I Wish I Missed My Ex          66   

    duration_ms  explicit         artist                  id_artists  \
0        354320         0          Queen  ['1dfeR4HaWDbWqFHLkxsg1d']   
1        252960         0           ABBA  ['0LcJLqbBmaGUft1e9Mm8HV']   
2        258773         0          Oasis  ['2DaxqgrOhkeH0fpeiQq2f4']   
3        266773         0       Coldplay  ['4gzpq5DPGxSnKTe4SA8HAU']   
4        270507         1        Outkast  ['1G9G7WwrXka3Z1r7aIDjI7']   
..          ...       ...            ...                         ...   
63       143667         1  Summer Walker  ['57LYzLEk2LcFghVwuWbcuS']   
64       113684         0    khai dreams  ['0lawSNBxNgJFQYJnQzLH8c']   
65       180019         0           Joji  ['3MZsBdqDrRTJihTHQrO6Dq']   
66       196800         1           ODIE  ['5aMIbwZQvP2MHPMVC5zCGj']   
67       203067         0        Mahalia  ['16rCzZOMQX7P8Kmn5YKexI']   

   release_date  danceability  energy  ...  loudness  mode  speechiness  \
0    1975-11-21         0.414   0.404  ...    -9.928     0       0.0499   
1          1976         0.354   0.535  ...    -8.876     1       0.0303   
2    1995-10-02         0.378   0.896  ...    -4.066     1       0.0407   
3    2000-07-10         0.429   0.661  ...    -7.227     1       0.0281   
4    2000-10-31         0.843   0.806  ...    -5.946     0       0.2690   
..          ...           ...     ...  ...       ...   ...          ...   
63   2019-10-04         0.759   0.413  ...    -7.282     0       0.0494   
64   2017-12-04         0.851   0.424  ...    -9.307     1       0.0463   
65   2019-06-14         0.552   0.650  ...    -7.199     1       0.1280   
66   2018-04-06         0.802   0.382  ...    -7.356     1       0.0312   
67   2019-09-06         0.669   0.669  ...    -4.185     1       0.0850   

    acousticness  instrumentalness  liveness  valence    tempo  \
0       0.271000          0.000000    0.3000    0.224   71.105   
1       0.627000          0.000003    0.0808    0.434  110.821   
2       0.000452          0.000000    0.1940    0.469  174.302   
3       0.002390          0.000121    0.2340    0.285  173.372   
4       0.143000          0.000000    0.0771    0.613   94.948   
..           ...               ...       ...      ...      ...   
63      0.700000          0.000000    0.1400    0.403  115.893   
64      0.383000          0.000000    0.0843    0.591   94.927   
65      0.422000          0.000275    0.3720    0.316  167.788   
66      0.792000          0.163000    0.0783    0.581   99.969   
67      0.274000          0.000000    0.2890    0.546   95.023   

    time_signature  favourite  
0                4        1.0  
1                4        1.0  
2                4        1.0  
3                4        1.0  
4                4        1.0  
..             ...        ...  
63               4        1.0  
64               4        1.0  
65               4        1.0  
66               4        1.0  
67               4        1.0  

[68 rows x 21 columns]

#### Of the 191 tracks in fav, only 68 corresponding audio features are present in favFeatures.  

In [5]:
#unqiue tracks in fav 
fav_track1 = fav.track.unique()
fav_track1[:15] #first 15 tracks

array(['Purple Emoji (feat. J. Cole)',
       'Sacrifices (with EARTHGANG & J. Cole feat. Smino & Saba)',
       'everybody dies', 'Change', 't h e . c l i m b . b a c k',
       'Pretty Little Fears (feat. J. Cole)',
       'Come Through and Chill (feat. J. Cole & Salaam Remi)',
       'Kevin’s Heart', 'Zendaya (feat. J. Cole)', 'Love Yourz',
       'Power Trip (feat. Miguel)', '4 Your Eyez Only', 'Adore You',
       'For Whom The Bell Tolls', 'MIDDLE CHILD'], dtype=object)

In [6]:
#unqiue tracks in favFeatures  
fav_track2 = favFeatures.track.unique()
fav_track2[:15] #first 15 tracks

array(['Bohemian Rhapsody - Remastered 2011', 'Fernando',
       'Wonderwall - Remastered', 'Yellow', 'Ms. Jackson', 'Fix You',
       "Sugar, We're Goin Down", "Day 'N' Nite (Nightmare)", 'Work Out',
       'The A Team', 'Lights Please', 'Thinkin Bout You',
       'Bitch, Don’t Kill My Vibe', 'Sweater Weather', 'No Role Modelz'],
      dtype=object)

In [7]:
missingTracks = dict()

for track in fav_track1:
    if track not in fav_track2:
        #using spotipy to search track name to get track id
        track_id = sp.search(q=' track:' + track, type='track')
        try:
            #populate missingTraks with each track name: track id
            missingTracks[track] = track_id['tracks']['items'][0]['id']
        except IndexError:
            print("Unable to find id for: " + track)

Unable to find id for: Honesty (Remix) [feat. Jessie Reyez]
Unable to find id for: Don't Check On Me (feat. Justin Bieber & Ink)
Unable to find id for: I'm the One (feat. Justin Bieber, Quavo, Chance the Rapper & Lil Wayne)


In [8]:
#make missingTracks dictionary into dataframe called ids
ids = pd.DataFrame(missingTracks.items(), columns=['track', 'id'])
ids.head()

track                      id
0                       Purple Emoji (feat. J. Cole)  3OJ4LEE50iFELaHiZ09zif
1  Sacrifices (with EARTHGANG & J. Cole feat. Smi...  7wTA0NKIm6T7nP2kaymU2a
2                                             Change  0I1eFRytp4XRhLCjT6tZm7
3                        t h e . c l i m b . b a c k  5lLNBIyjp72btcnrjBG751
4                Pretty Little Fears (feat. J. Cole)  4at3d5QWnlibMVN75ECDrp

#### Unable to find track id for some songs

Since there is only 3 songs, I manually found their respective track ids.

- Honesty (Remix) [feat. Jessie Reyez] : 47r6MeqPoeQgCsa5vOQxoo
- Don't Check On Me (feat. Justin Bieber & Ink) : 7ILEWkpfdK6AjH2D7jppWk
- I'm the One (feat. Justin Bieber, Quavo, Chance the Rapper & Lil Wayne) : 3DXncPQOG4VBw3QHh3S817

In [9]:
track1 = pd.Series(data={"track":"Honesty (Remix) [feat. Jessie Reyez]", "id":"47r6MeqPoeQgCsa5vOQxoo"})

In [10]:
track2 = pd.Series(data={"track":"Don't Check On Me (feat. Justin Bieber & Ink)", "id":"7ILEWkpfdK6AjH2D7jppWk"})

In [11]:
track3 = pd.Series(data={"track":"I'm the One (feat. Justin Bieber, Quavo, Chance the Rapper & Lil Wayne)", "id":"3DXncPQOG4VBw3QHh3S817"})

In [12]:
#append the 3 tracks
ids = ids.append(track1, ignore_index=True)
ids = ids.append(track2, ignore_index=True)
ids = ids.append(track3, ignore_index=True)
ids.tail()

track                      id
118                                         Be Alright  3EPXxR3ImUwfayaurPi3cm
119                    Beautiful People (feat. Khalid)  4evmHXcjt3bTUHD1cvny97
120               Honesty (Remix) [feat. Jessie Reyez]  47r6MeqPoeQgCsa5vOQxoo
121      Don't Check On Me (feat. Justin Bieber & Ink)  7ILEWkpfdK6AjH2D7jppWk
122  I'm the One (feat. Justin Bieber, Quavo, Chanc...  3DXncPQOG4VBw3QHh3S817

In [13]:
def get_features(df):
    playlist_features_list = ["id", "track", "duration_ms", "danceability","energy","key","loudness","mode", "speechiness","acousticness", "instrumentalness","liveness","valence","tempo", "time_signature"]
    playlist_df = pd.DataFrame(columns = playlist_features_list)
    
    for i in range(len(df)): 
        playlist_features = {} 
        playlist_features['id'] = df['id'][i]
        playlist_features['track'] = df['track'][i]
        playlist_features['favourite'] = np.int32(1)
        
        audio_features = sp.audio_features(playlist_features['id'])[0]

        for feature in playlist_features_list[2:]:
            playlist_features[feature] = audio_features[feature]

        
        track_df = pd.DataFrame(playlist_features, index = [0])
        playlist_df = pd.concat([playlist_df, track_df], ignore_index = True)
    
    return playlist_df

In [14]:
#dataframe with audio features of songs that were missing from favFeatures
missingFeatures = get_features(ids)
missingFeatures

id  \
0    3OJ4LEE50iFELaHiZ09zif   
1    7wTA0NKIm6T7nP2kaymU2a   
2    0I1eFRytp4XRhLCjT6tZm7   
3    5lLNBIyjp72btcnrjBG751   
4    4at3d5QWnlibMVN75ECDrp   
..                      ...   
118  3EPXxR3ImUwfayaurPi3cm   
119  4evmHXcjt3bTUHD1cvny97   
120  47r6MeqPoeQgCsa5vOQxoo   
121  7ILEWkpfdK6AjH2D7jppWk   
122  3DXncPQOG4VBw3QHh3S817   

                                                 track duration_ms  \
0                         Purple Emoji (feat. J. Cole)      163755   
1    Sacrifices (with EARTHGANG & J. Cole feat. Smi...      382307   
2                                               Change      198213   
3                          t h e . c l i m b . b a c k      306257   
4                  Pretty Little Fears (feat. J. Cole)      240342   
..                                                 ...         ...   
118                                         Be Alright      196373   
119                    Beautiful People (feat. Khalid)      197867   
120               Honesty (Remix) [feat. Jessie Reyez]      187805   
121      Don't Check On Me (feat. Justin Bieber & Ink)      204880   
122  I'm the One (feat. Justin Bieber, Quavo, Chanc...      288600   

     danceability  energy key  loudness mode  speechiness  acousticness  \
0           0.707   0.543   8    -4.610    1       0.2790        0.1510   
1           0.825   0.672   0    -7.108    1       0.4180        0.0858   
2           0.247   0.438   5    -8.479    1       0.0358        0.0455   
3           0.618   0.632  11    -6.973    1       0.3460        0.0526   
4           0.610   0.458   9   -11.336    1       0.3600        0.6960   
..            ...     ...  ..       ...  ...          ...           ...   
118         0.553   0.586  11    -6.319    1       0.0362        0.6970   
119         0.640   0.648   5    -8.113    0       0.1870        0.1240   
120         0.725   0.374   6    -6.501    0       0.0395        0.7730   
121         0.661   0.395   7    -5.242    1       0.0333        0.8730   
122         0.609   0.668   7    -4.284    1       0.0367        0.0552   

    instrumentalness  liveness  valence    tempo time_signature  favourite  
0                  0    0.3840    0.824  144.180              4        1.0  
1                  0    0.1230    0.246  147.010              4        1.0  
2           0.000871    0.1110    0.390  185.716              3        1.0  
3                  0    0.3530    0.319   81.446              4        1.0  
4           0.000103    0.0852    0.246  192.099              4        1.0  
..               ...       ...      ...      ...            ...        ...  
118                0    0.0813    0.443  126.684              4        1.0  
119                0    0.0802    0.548   92.977              4        1.0  
120                0    0.1110    0.344   80.032              4        1.0  
121                0    0.1120    0.679   82.028              4        1.0  
122                0    0.1670    0.811   80.924              4        1.0  

[123 rows x 16 columns]

In [15]:
#save to csv
missingFeatures.to_csv('missingFeatures.csv', index=False)